Keep ICC profile when saving image, if present

siovene opened the pull request:
Closes #366

cogat commented on easy_thumbnails/engine.py:
```diff
@@ -57,6 +62,9 @@ def save_image(image, destination=None, filename=None, **options):
                 max(image.size) >= settings.THUMBNAIL_PROGRESSIVE):
             options['progressive'] = True
         try:
+            icc_profile = image.info.get('icc_profile')
```
I wonder how expensive this is? Perhaps put it inside `if 'keep_icc_profile' in thumbnail_options:` so that it doesn't run unless needed (in case it is expensive)

cogat commented on easy_thumbnails/engine.py:
```diff
@@ -35,10 +36,14 @@ def process_image(source, processor_options, processors=None):
+    options = {
+        'quality': thumbnail_options.get('quality', 100),
+        'subsampling': thumbnail_options.get('subsampling', 2)
+    }
```
Were these defined somewhere else? If so I'd expect to be seeing changes in those lines as part of the PR.

cogat commented on easy_thumbnails/engine.py:
```diff
@@ -57,6 +62,9 @@ def save_image(image, destination=None, filename=None, **options):
             options['progressive'] = True
         try:
+            icc_profile = image.info.get('icc_profile')
+            if icc_profile and 'keep_icc_profile' in thumbnail_options:
```
I think change to `thumbnail_options.get('keep_icc_profile')` - the value might be set to `False`.

cogat commented on easy_thumbnails/engine.py:
```diff
@@ -35,10 +36,14 @@ def process_image(source, processor_options, processors=None):
 
 
-def save_image(image, destination=None, filename=None, **options):
+def save_image(image, thumbnail_options, destination=None, filename=None):
```
It seems a shame to lose `**options` here. There may be other implementations out in the world that use it. Couldn't you add the `keep_icc_profile` as a kwarg?

cogat commented on easy_thumbnails/tests/test_engine.py:
```diff
@@ -8,12 +8,12 @@ class SaveTest(TestCase):
     def test_save_jpeg_rgba(self):
         source = Image.new('RGBA', (100, 100), (255, 255, 255, 0))
-        data = engine.save_image(source, filename='test.jpg')
+        data = engine.save_image(source, {}, filename='test.jpg')
```
keeping `**options` above would avoid the need to change the call signature.

cogat commented on easy_thumbnails/engine.py:
```diff
@@ -57,6 +62,9 @@ def save_image(image, destination=None, filename=None, **options):
+            icc_profile = image.info.get('icc_profile')
+            if icc_profile and 'keep_icc_profile' in thumbnail_options:
+                options['icc_profile'] = icc_profile
             image.save(destination, format=format, optimize=1, **options)
```
Perhaps add a test that ensures that the option produces the expected outcome??

cogat reviewed (changes requested):
(I don't have a commit bit so may be unqualified).

This is a useful feature, quite neatly implemented. I would like to see if it can be done without altering the signature of `save_image` and hence needing to alter the tests.

siovene commented:
Hi @cogat, thanks for the review! I think I addressed all points. They were very valid and it looks a lot better now :) Can you give it another look, please?

Sorry for the number of commits, I assume there will be a squash & merge at the end.

Thanks!

siovene commented:
Just for piece of mind, this PR has been in production on AstroBin for half a year with no problems :)

siovene commented:
@SmileyChris any chance this could be reviewed? It's the only thing that keeps me from using the upstream library, and it seems to work fine: it's been in production on AstroBin for half a year, which is a volume of about 50k images. Thanks!

BigglesZX commented:
Any chance of getting this merged in?

BigglesZX commented:
@SmileyChris @jrief Is it possible this could be reviewed and merged in? I've been using a fork based on this PR for a couple of years in production with no issues.

SmileyChris commented:
From my fading memory, here's some context as to why this wasn't merged in earlier.

I believe we did actually keep the icc profile at some stage, but it turned out that some images can have incredibly large ICC profiles attached so this blew the size of the thumbnails out. The easy solution was to remove the profile. If possible, it would be nice to only include the profile if it was smaller than X, assuming finding this out is doable...

SmileyChris commented:
Since this is an option which defaults to false however, I can't see any harm in merging it! Maybe the docs could mention the side effect of this producing potentially larger images. A reasonable next step too would be that when this option is false, actually squash the color space before dropping the icc profile...
